Feat:  Admin should be able to set roles and permissions #184361216

ISHIMWE-Aime opened the pull request:
# Description
I have add the feature where admin should set roles and permission;
Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

# Type of change

Please check the options that are relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature or Enhancement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project
- [ ] I have created a loom video for the new feature or enhancement
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

emilenfc reviewed (dismissed):
LGTM

shyakadavis commented on src/routes/roleAndPermission.routes.ts:
```diff
@@ -0,0 +1,50 @@
+
+setRoleAndPermissionRouter.patch(
+  "/update/:id",
+  ValidateJoi(RoleAndPermissionsSchema.roleTobeSet),
+  [protectRoute, isAdmin],
```
```suggestion
  [protectRoute, isAdmin],
  ValidateJoi(RoleAndPermissionsSchema.roleTobeSet)
```

shyakadavis commented on src/routes/roleAndPermission.routes.ts:
```diff
@@ -0,0 +1,50 @@
+
+setRoleAndPermissionRouter.patch(
+  "/permissions/add/:role",
+  ValidateJoi(RoleAndPermissionsSchema.permissionsToAdd),
+  [protectRoute, isAdmin],
```
```suggestion
  [protectRoute, isAdmin],
  ValidateJoi(RoleAndPermissionsSchema.permissionsToAdd),
```

shyakadavis commented on src/routes/roleAndPermission.routes.ts:
```diff
@@ -0,0 +1,50 @@
+//   ,
+setRoleAndPermissionRouter.delete(
+  "/permissions/remove/:role",
+  ValidateJoi(RoleAndPermissionsSchema.permissionsToRemove),
+  [protectRoute, isAdmin],
```
```suggestion
  [protectRoute, isAdmin],
  ValidateJoi(RoleAndPermissionsSchema.permissionsToRemove),
```

shyakadavis commented on src/routes/roleAndPermission.routes.ts:
```diff
@@ -0,0 +1,50 @@
+
+setRoleAndPermissionRouter.post(
+  "/set",
+  ValidateJoi(RoleAndPermissionsSchema.roleAndPermissions),
+  [protectRoute, isAdmin],
```
```suggestion
  [protectRoute, isAdmin],
  ValidateJoi(RoleAndPermissionsSchema.roleAndPermissions),
```

shyakadavis commented on src/middleware/auth/auth.middleware.ts:
```diff
@@ -93,6 +93,7 @@ export const isAdmin = async (
     }
   } catch (error) {
     if (error instanceof Error) {
+      console.log("I s runnign !!");
```
```suggestion
```

shyakadavis commented on src/database/seeders/20230314055804-RoleAndPermision.js:
```diff
@@ -0,0 +1,55 @@
+            "Set permissions",
+            "Update product",
+            "Add product",
+            "View oders",
```
```suggestion
            "View orders",
```

shyakadavis commented on src/database/seeders/20230314055804-RoleAndPermision.js:
```diff
@@ -0,0 +1,55 @@
+            "Set permissions",
+            "Update product",
+            "Add product",
+            "View oders",
```
```suggestion
            "View orders",
```

musllim commented on src/controllers/RoleAndPermission.controller.ts:
```diff
@@ -0,0 +1,228 @@
+export const addPermissions = async (req: Request, res: Response) => {
+  try {
+    const role = req.params.role;
+    const permissions: [] = req.body.permissions;
```
add type of those array this defaults it to never

musllim commented on src/controllers/RoleAndPermission.controller.ts:
```diff
@@ -0,0 +1,228 @@
+          permissions[i],
+        );
+        if (index !== -1) {
+          alreadyExist.push(permissions[i]);
```
How is this possible? You will be getting errors like this ```Argument of type 'string' is not assignable to parameter of type 'never'```

musllim reviewed (commented):
Look at that comments

mucyoalexaxel commented:
@ISHIMWE-Aime I think you missed the whole point of how to implement this feature I would suggest you to rethink the logic and make enough research of how to implement this.

SergeInGithub reviewed (commented):
Checking through this, I noticed that you didn't address the suggested changes @shyakadavis gave you. Can you please address them?

mucyoalexaxel commented on src/controllers/Auth.controller.ts:
@ISHIMWE-Aime refactor this file, below are key things you should consider while refactoring.
1. Destructure `dataValues` for better readability.
2. Use consistent status codes for errors.
3. Use async/await with try-catch blocks for better error handling.
4. fix hound errors

mucyoalexaxel commented on src/controllers/Notifications.controller.ts:
```diff
@@ -11,6 +11,9 @@ import { NotificationAttributes } from "../interfaces";
 import { verifyToken } from "../utils";
 import User from "../database/models/User.model";
 
+const B_ACCESSKEY = process.env.B_ACCESSKEY as string;
+const S_ACCESSKEY = process.env.S_ACCESSKEY as string;
```
```suggestion
import { B_ACCESSKEY, S_ACCESSKEY } from "../utils"
```
create a role constants files in utils and declare the role constants there instead of storing them in your .env

mucyoalexaxel commented on src/controllers/Role.controller.ts:
```diff
@@ -0,0 +1,132 @@
+};
+
+// Admin set roles in db
+export const setRole = async (req: Request, res: Response) => {
```
```suggestion
export const createRole = async (req: Request, res: Response) => {
```

mucyoalexaxel commented on src/controllers/Role.controller.ts:
```diff
@@ -0,0 +1,132 @@
+  }
+};
+
+// Admin set roles in db
```
```suggestion
```

mucyoalexaxel commented on src/controllers/Role.controller.ts:
```diff
@@ -0,0 +1,132 @@
+  }
+};
+
+// Admin set role to a particular user
```
```suggestion
```

mucyoalexaxel commented on src/controllers/Role.controller.ts:
```diff
@@ -0,0 +1,132 @@
+};
+
+// Admin set role to a particular user
+export const setRoleToUser = async (req: Request, res: Response) => {
```
```suggestion
export const setRoleByUserId = async (req: Request, res: Response) => {
```

mucyoalexaxel commented on src/controllers/Role.controller.ts:
```diff
@@ -0,0 +1,132 @@
+
+// Admin set role to a particular user
+export const setRoleToUser = async (req: Request, res: Response) => {
+  try {
+    const userId = req.params.id;
+
+    if (isValidUuid(req.body.roleId)) {
+      const updatedUserRole = await User.update(
+        { roleId: req.body.roleId },
+        {
+          where: { id: userId },
+          returning: true,
+        },
+      );
+
+      return res.status(200).json({
+        status: 200,
+        success: true,
+        data: updatedUserRole,
+      });
+    } else {
+      return res.status(400).json({
+        status: 400,
+        success: false,
+        message: "Please provide a valid uuid for roleId",
+      });
+    }
+  } catch (error) {
+    if (error instanceof Error) {
+      return res.status(500).json({
+        status: 500,
+        success: false,
+        message: "Error occur while giving role a specified user",
+        error: error.message,
+      });
+    }
+  }
+};
```
@ISHIMWE-Aime refactor this code block

ISHIMWE-Aime replied:
I am sorry but I am not getting clearly what you mean here.
Thank you.

mucyoalexaxel commented on src/controllers/Role.controller.ts:
```diff
@@ -0,0 +1,132 @@
+};
+
+// delete a specific role
+export const deleteARole = async (req: Request, res: Response) => {
+  try {
+    if (isValidUuid(req.params.id)) {
+      const deletationRes = await deleteRoleService(req.params.id);
+
+      if (deletationRes === 0) {
+        return res.status(404).json({
+          status: 404,
+          success: false,
+          message: "Role doesn't exist!",
+        });
+      } else {
+        return res.status(200).json({
+          status: 200,
+          success: true,
+          message: "Role deleted successfully",
+          data: deletationRes,
+        });
+      }
+    } else {
+      return res.status(400).json({
+        status: 400,
+        success: false,
+        message: "Please provide a valid uuid for Role to delete",
+      });
+    }
+  } catch (error) {
+    if (error instanceof Error) {
+      return res.status(500).json({
+        status: 500,
+        success: false,
+        message: "Error occur while deleting role",
+        error: error.message,
+      });
+    }
+  }
+};
```
@ISHIMWE-Aime I think you missed something here what if you delete a role which is already assigned to a user what happens ?

ISHIMWE-Aime replied:
The role will be deleted with its associations in users table and if the role is seller, the associations in product table will be deleted as well.
That all will happen. 
If the there is anything wrong please help me to understand.

mucyoalexaxel commented on src/database/migrations/20230227054250-create_users_table.js:
```diff
@@ -100,6 +106,16 @@ module.exports = {
         allowNull: true,
       },
     });
+    // create a relationship between users and roleAndPermissions
```
```suggestion
```

mucyoalexaxel commented on src/middleware/auth/auth.middleware.ts:
@ISHIMWE-Aime have you reviewed this file

ISHIMWE-Aime replied:
Thank you for your review but I am not getting exactly what you want to mean @mucyoalexaxel 
Apart from that what I have just done is to made one middleware function to check if user is authenticated and check a permission. 
that middleware is located in the file called authanticated.middleware.ts.

ISHIMWE-Aime replied:
That will remove those three middleware functions "isAuth", "isAdmin" and "isSeller" as you had suggested.  
Thank you; if there is any refinement need, I am willing to hear from you.

mucyoalexaxel commented on src/middleware/auth/authanticated.middleware.ts:
@ISHIMWE-Aime refactor this file as well

mucyoalexaxel reviewed (changes requested):
@ISHIMWE-Aime kindly check the suggested changes refactor all the required components, fix build errors and tests as well

shyakadavis commented:
@ISHIMWE-Aime  I was going through the code, and noticed you didn't address 2 major issues.
The suggestions from Alex, and the hound errors.

May you please first go over them, so that we can get room to review after the initial requests have been addressed.

Thanks.

musllim committed:
ft: admin set roles

fx: fix github actions tests failing reporting and coveralls

we well getting green check mark indicating that tests are passing, though they were failing,
this fix will stop coveralls reporting and will give red failing cross on github actions if tests are fails.

Fixes #123

add chatting feature for users #184361239

add peer to peer chatting
remove broadcasting message to all users

fx: redirect user to the deployed frontend login page #121

after successfully being registeresd , when user click the verification link in the email it should redirect to the deployed frontend login page

Fixes #121

fx: add google login middleware

This app had no way to logging using google, the client needed this feature which will return the response to client.

ft: admin set roles

fx: redirect user to the deployed frontend login page #121

after successfully being registeresd , when user click the verification link in the email it should redirect to the deployed frontend login page

Fixes #121

fx: add google login middleware

This app had no way to logging using google, the client needed this feature which will return the response to client.

ft: admin set roles and permissions

fx: add google login middleware

This app had no way to logging using google, the client needed this feature which will return the response to client.

ft: admin set roles and permissions

Revert "add ft: admin should be able to disable any user account #184361215"

add chatting feature for users #184361239

ft-buyer-provide-review#184361226

fx: fix github actions tests failing reporting and coveralls

we well getting green check mark indicating that tests are passing, though they were failing,
this fix will stop coveralls reporting and will give red failing cross on github actions if tests are fails.

Fixes #123

 ft: admin set roles and permissions

ft-buyer-provide-review#184361226

fx: fix github actions tests failing reporting and coveralls

we well getting green check mark indicating that tests are passing, though they were failing,
this fix will stop coveralls reporting and will give red failing cross on github actions if tests are fails.

Fixes #123

ft: admin set roles and permissions

ft: admin set roles and permissions

add ft: admin should be able to disable any user account
add documentation and unit testing for this feature

WHEN a user is suspected or proved to have violated the terms of service of the application
THEN admin is able to trigger an action to block usage of a particular account and an email sent to the user with reasons fro the block.

Revert "add ft: admin should be able to disable any user account #184361215"

add chatting feature for users #184361239

ft-buyer-provide-review#184361226

ft: admin set roles and permissions

ft(notification): read all/a notification by ID [#184361235]

fx:change the way to get orders #184361232

ft: admin set roles and permissions

fx(rebase): to be up to date with upstream

Fix(Hound CI) fixing houndCI Error

ft:seller delete specific Item in collection

ft: #184361213 add Two Factor Authentication for Sellers

add all tests for this feature plus it's documentation

GIVEN there is a request to log in a user to the website
WHEN 2FA is enabled on the particular user
THEN Further request for a 2FA token should be requested and once the token is received and validated, the user is allowed to log in.

bg: won't treat an empty cart as an error & remove the required middleware to add

fx(rebase): to be up to date with upstream

ft: #184361213 add Two Factor Authentication for Sellers

add all tests for this feature plus it's documentation

GIVEN there is a request to log in a user to the website
WHEN 2FA is enabled on the particular user
THEN Further request for a 2FA token should be requested and once the token is received and validated, the user is allowed to log in.

bg: won't treat an empty cart as an error & remove the required middleware to add

fx(rebase): to be up to date with upstream

ft: admin set roles and permissions

ft: admin set roles and permissions

Revert "add ft: admin should be able to disable any user account #184361215"

add chatting feature for users #184361239

ft-buyer-provide-review#184361226

fx: fix github actions tests failing reporting and coveralls

we well getting green check mark indicating that tests are passing, though they were failing,
this fix will stop coveralls reporting and will give red failing cross on github actions if tests are fails.

Fixes #123

 ft: admin set roles and permissions

ft-buyer-provide-review#184361226

fx: fix github actions tests failing reporting and coveralls

we well getting green check mark indicating that tests are passing, though they were failing,
this fix will stop coveralls reporting and will give red failing cross on github actions if tests are fails.

Fixes #123

ft: admin set roles and permissions

ft: admin set roles and permissions

add ft: admin should be able to disable any user account
add documentation and unit testing for this feature

WHEN a user is suspected or proved to have violated the terms of service of the application
THEN admin is able to trigger an action to block usage of a particular account and an email sent to the user with reasons fro the block.

Revert "add ft: admin should be able to disable any user account #184361215"

add chatting feature for users #184361239

ft-buyer-provide-review#184361226

ft: admin set roles and permissions

ft(notification): read all/a notification by ID [#184361235]

fx:change the way to get orders #184361232

ft: admin set roles and permissions

fx(rebase): to be up to date with upstream

Fix(Hound CI) fixing houndCI Error

ft:seller delete specific Item in collection

ft: #184361213 add Two Factor Authentication for Sellers

add all tests for this feature plus it's documentation

GIVEN there is a request to log in a user to the website
WHEN 2FA is enabled on the particular user
THEN Further request for a 2FA token should be requested and once the token is received and validated, the user is allowed to log in.

bg: won't treat an empty cart as an error & remove the required middleware to add

fx(rebase): to be up to date with upstream

ft: #184361213 add Two Factor Authentication for Sellers

add all tests for this feature plus it's documentation

GIVEN there is a request to log in a user to the website
WHEN 2FA is enabled on the particular user
THEN Further request for a 2FA token should be requested and once the token is received and validated, the user is allowed to log in.

bg: won't treat an empty cart as an error & remove the required middleware to add

fx(rebase): to be up to date with upstream

ft: admin set roles and permissions

emilenfc commented on src/routes/wishes.routes.ts:
```diff
@@ -5,16 +5,25 @@ import {
 } from "../controllers";
-import { isAuth } from "../middleware/auth";
+import { AuthAndRoleChecker } from "../middleware/auth/authanticated.middleware";
+import { B_ACCESSKEY } from "../utils/roleConstants";
```
I want to point out that you are currently using hardcoded roles `B_ACCESSKEY, A_ACCESSKEY, and S_ACCESSKEY.` However, I don't think this will work properly if the admin assigns a role ID to a user that is different from these hardcoded ones.

ISHIMWE-Aime replied:
Thank you @emilenfc ,  
but If you have just checked the suggested changes from @mucyoalexaxel, he has just suggested that instead of using environmental variable.

emilenfc reviewed (commented):
Hi @ISHIMWE-Aime, great job on the work you've done so far! However, I noticed something that I'd like to clarify. It seems like you're using Roles in a way that doesn't require creating new roles. If I'm mistaken, please let me know. And since you're not creating new roles, there should be no need to have functions for that. I also noticed that you haven't made the changes requested by @mucyoalexaxel  yet, so please work on those. Finally, please provide tests for this task as well.

musllim commented on test/updatePassword.spec.ts:
```diff
@@ -20,7 +20,10 @@ describe("update password test", () => {
     };
 
     const createUser = await request(app).post("/api/v1/user").send(user);
+    console.log(createUser, "the createUser is !!!!!!!!!!!!!");
+    console.log(createUser.body.data, "The confrim source is :::::");
```
remove these lines

shyakadavis commented on src/middleware/auth/authanticated.middleware.ts:
```diff
@@ -0,0 +1,70 @@
+            return res.status(401).json({
+              status: 401,
+              success: false,
+              message: "You are not allowed to perform this action!",
```
```suggestion
              message: "Unauthorized",
```

shyakadavis commented on docs/roleAndPermission.docs.yaml:
```diff
@@ -0,0 +1,134 @@
+paths:
```
Refactor this documentation to reflect the actual routes. Take a closer look. This documentation is out of date with the routes files. Also, you can rename this file to `role.docs.yaml` to maintain consistency.

shyakadavis commented on test/auth.spec.ts:
```diff
@@ -1,12 +1,12 @@
-    await sequelize.truncate({ cascade: true }); // deletes all data from all tables
-    await sequelize.close(); // closes the connection to the database
-  });
+  // afterAll(async () => {
+  //   await sequelize.truncate({ cascade: true }); // deletes all data from all tables
+  //   await sequelize.close(); // closes the connection to the database
+  // });
```
```suggestion

```

shyakadavis commented on test/user.spec.ts:
```diff
@@ -1,6 +1,6 @@
 import request from "supertest";
 import app from "../src/app";
-import { sequelize } from "../src/database/models";
+// import { sequelize } from "../src/database/models";
```
```suggestion
```

shyakadavis commented on test/user.spec.ts:
```diff
@@ -17,10 +17,10 @@ describe("🧑‍🤝‍🧑 USERS UNIT", () => {
-    await sequelize.truncate({ cascade: true }); // deletes all data from all tables
-    await sequelize.close(); // closes the connection to the database
-  });
+  // afterAll(async () => {
+  //   await sequelize.truncate({ cascade: true }); // deletes all data from all tables
+  //   await sequelize.close(); // closes the connection to the database
+  // });
```
```suggestion
```

shyakadavis commented on tsconfig.json:
```diff
@@ -11,6 +11,12 @@
+    "docs/swagger.ts",
+    "test",
+    "docs",
+    "src/middleware/permissionCheck"
```
```suggestion

```

I think you can safely remove this because no such file exists in the code base.

shyakadavis reviewed (changes requested):
Awesome work @ISHIMWE-Aime 


Please take a look at the comments, and the rest should look good to me. 🚀

emilenfc reviewed (changes requested):
@ISHIMWE-Aime  You did Nice work, but make sure you Rebase first of all

mucyoalexaxel reviewed (commented):
@shyakadavis @emilenfc @musllim @SergeInGithub . 
Review This PR smoothly

shyakadavis commented on src/controllers/Auth.controller.ts:
```diff
@@ -25,10 +25,52 @@ export const loginUser = async (req: Request, res: Response) => {
 
     const pwdMatch = await bcrypt.compare(password, user.dataValues.password);
     if (!pwdMatch) throw new Error(loginError);
+    const {
+      id,
+      twoFAenabled,
+      twoFAverified,
+      secret,
+      surname,
+      role,
+      status,
+      confirmationCode,
+      googleId,
+      resetToken,
+      province,
+      district,
+      sector,
+      cell,
+      street,
+      avatar,
+      lastPasswordUpdate,
+    } = user;
```
Personally, I think this is a disaster.

You can't encode all that info inside a token without exposing yourself to security vulnerabilities.

emilenfc replied:
Based on the request of this task on pivotaltracker, I think the way this task is implemented is right. Also storing(sending) all this data in a token can be fine and easy to acces, but I'm not sure about the security of the user data as we may send this token in cookie or local storage on FE

shyakadavis commented on src/controllers/Auth.controller.ts:
```diff
@@ -114,7 +156,48 @@ export const resetPasswordRequestController = async (
+        avatar,
+        lastPasswordUpdate,
+      },
+      "5m",
```
@ISHIMWE-Aime 

There is a concern for this.

A short expiration date for tokens is more secure. But there is a catch.

We don't have `refresh-tokens` set in place. So, this means that a user will be forced to log in every 5 minutes. This is a terrible experience for the user.

I would suggest to increase this time, because we don't have refresh tokens set up, and we are out of time to implement that.

shyakadavis reviewed (changes requested):
@ISHIMWE-Aime 

Please address some concerns expressed below.

musllim commented on src/controllers/Auth.controller.ts:
Try to refactor this file, and avoid too many nesting of if statements.

SergeInGithub commented on src/middleware/auth/authanticated.middleware.ts:
```diff
@@ -0,0 +1,90 @@
+        try {
+          const token = req.headers.authorization.split(" ")[1];
+          const {
+            id,
```
the same applies here, you're verifying this encoded data but you shouldn't encode this much data in the first place

emilenfc reviewed (changes requested):
@ISHIMWE-Aime  You did a great things so far ✅, but also check on the requested changes

SergeInGithub reviewed (commented):
@ISHIMWE-Aime check the reviews i left

emilenfc commented on docs/role.docs.yaml:
@ISHIMWE-Aime  remove those unused comments in this file

shyakadavis replied:
@ISHIMWE-Aime

emilenfc commented on docs/roleAndPermission.docs.yaml:
@ISHIMWE-Aime  Remove also this file as it is no longer needed

emilenfc reviewed (changes requested):
@ISHIMWE-Aime great work, however first of all trie check the request I left

shyakadavis commented on src/controllers/Auth.controller.ts:
```diff
@@ -114,7 +134,7 @@ export const resetPasswordRequestController = async (
       .json({ status: 404, success: false, message: "Email does not exist" });
   } else {
-    const resetToken = await generateToken(user.dataValues.email, "5m");
+    const resetToken = await generateToken(user.dataValues, "5m");
```
I think this is contrary to what we discussed.

I don't think you are supposed to sign the entire user's data inside the token.

Just the `email` or `id` is enough

shyakadavis replied:
You are still setting the entire user object on the token? Is there a specific reason why you are doing this?

shyakadavis reviewed (commented):
Please take at some minor changes to do with with where you are generating the token. I think you missed a few places.

shyakadavis commented on docs/role.docs.yaml:
```diff
@@ -0,0 +1,252 @@
+paths:
+  # Admin set roles
+  /api/v1/role:
```
You forgot to resolve the lines commented in this file.

shyakadavis commented on docs/roleAndPermission.docs.yaml:
```diff
@@ -0,0 +1,134 @@
+# paths:
```
And this as well.

shyakadavis commented on src/services/protectRoutes.service.ts:
```diff
@@ -1,5 +1,5 @@
 import { NextFunction, Request, Response } from "express";
-import * as JWT from "jsonwebtoken";
+// import * as JWT from "jsonwebtoken";
```
```suggestion
```

shyakadavis commented on src/services/protectRoutes.service.ts:
```diff
@@ -9,7 +9,7 @@ export const protectRoute = async (
 ) => {
   let token;
-  const secret = process.env.SECRET_TOKEN as string;
+  // const secret = process.env.SECRET_TOKEN as string;
```
```suggestion
```

shyakadavis commented on src/middleware/auth/authanticated.middleware.ts:
```diff
@@ -0,0 +1,63 @@
+            role,
+            avatar,
+          };
+          console.log(role, accessKey);
```
```suggestion
```

shyakadavis commented on src/database/models/User.model.ts:
```diff
@@ -127,6 +135,9 @@ const User = sequelize.define<UserInstance>(
   },
 );
 
+// Add the belongsTo association to the Product model
```
```suggestion
```

shyakadavis commented on src/controllers/User.controller.ts:
```diff
@@ -73,9 +80,7 @@ export const createUser = async (req: Request, res: Response) => {
-      console.log(`Error creating user: ${error.message}`);
-      console.log(error);
-
+      console.log(`Error creating user: ${error}`);
```
```suggestion
```

shyakadavis reviewed (changes requested):
Please take a look at the requested and suggested changes.

emilenfc commented on src/routes/user.routes.ts:
```diff
@@ -1,13 +1,31 @@
+userRouter.get(
+  "/:id",
+  AuthAndRoleChecker(() => {
+    return { value: ADMIN_ACCESSKEY };
```
to get one user is not for admin, what if user want to view account.
```suggestion

```

emilenfc commented on docs/role.docs.yaml:
```diff
@@ -0,0 +1,99 @@
+        "404":
+          description: Not Found
+
+  /api/v1/role/assign/{user_id}:
```
based on route you have 
```suggestion
  /api/v1/role/update/{user_id}:
```

musllim reviewed (dismissed):
LGTM
